ComponentKit/Core/Scope/CKComponentScopeTypes.h

```diff
@@ -16,3 +16,36 @@ typedef int32_t CKComponentScopeHandleIdentifier;
 typedef int32_t CKComponentScopeRootIdentifier;
 
 typedef std::unordered_multimap<CKComponentScopeHandleIdentifier, id (^)(id)> CKComponentStateUpdateMap;
+
+@protocol CKScopedComponent;
+@protocol CKScopedComponentController;
+
+/**
+ Enumerator blocks allow a consumer to enumerate over all of the components or controllers that matched a predicate.
+ */
+typedef void (^CKComponentScopeEnumerator)(id<CKScopedComponent>);
+typedef void (^CKComponentControllerScopeEnumerator)(id<CKScopedComponentController>);
+
+/**
+ Scope predicates are a tool used by the framework to register components and controllers on initialization that have
+ specific characteristics. These predicates allow rapid enumeration over matching components and controllers.
+ */
+typedef BOOL (*CKComponentScopePredicate)(id<CKScopedComponent>);
+typedef BOOL (*CKComponentControllerScopePredicate)(id<CKScopedComponentController>);
+
+
+@protocol CKComponentScopeEnumeratorProvider <NSObject>
```

Comment:
Could we decouple this even further? The fact that we are enumerating over the Scope Root is totally in the implementation detail. Could we rename this to something like: `CKComponentPredicateMatcher`?